Bring Plutus language version to the type level

lehins opened the pull request:
# Description

Fixes #3849 and #3406

Other improvements:

* Instead of accepting `UTxO` the validation function `ppViewHashesMatch` now accepts  `ScriptsProvided` and remove the no longer needed `languages` function. This approach avoids duplicate computation of `ScritpsProvided`, which was done in the `languages` function.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

lehins committed:
Cleanup haddock

lehins committed:
Rename `BinaryPlutus` to `PlutusBinary` for consistency

lehins committed:
Bring the Plutus Language version to the type level

This is a major change to how we represent Plutus Scripts. This affects
how we collect and evaluate plutus scripts. There is now a unified
interface for all plutus versions and incorrect usage of an unsupported
plutus version in all eras is prevented by the type system.

Translation of PlutusContext has also been overhauled. There is still
quite a bit duplication happens with TxInfo translation, but this will
be taken care of by a subsequent commit.

lehins committed:
Move PlutusScriptApi -> Plutus.Evaluate

lehins committed:
Move PlutusContext type classes into the new module:

* `Cardano.Ledger.Alonzo.Plutus.Context`

lehins committed:
Move plutus context type families from cardano-ledger-core

lehins committed:
Switch to ContextError handling using `Inject`

* Reduce a whole lot of duplication in TxInfo translation
* Changelog entries for all of the changes in #3889

teodanciu commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/AlonzoBBODY.hs:
These `Example` tests from this folder must have been a nightmare to adjust. :grimacing:

Soupstraw commented on eras/babbage/test-suite/test/Tests.hs:
```diff
@@ -25,7 +25,6 @@ defaultTests =
     [ Golden.tests
     , Tripping.tests
     , txInfoTests (Proxy @Babbage)
-    , txInfoTestsBabbageOnly
```
Why was this test removed?

lehins replied:
That is because the only thing it was testing was presence of incompatible Plutus versions within a translation, which is no longer possible thanks to the type checker. 

https://github.com/input-output-hk/cardano-ledger/blob/bcf6051f4a0b69c82e4de7327d7d8801fa1bad87/eras/babbage/test-suite/src/Test/Cardano/Ledger/Babbage/TxInfo.hs#L346C1-L351

Soupstraw commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/TreeDiff.hs:
```diff
@@ -29,6 +32,12 @@ instance ToExpr PoolVotingThresholds
 
 instance ToExpr DRepVotingThresholds
 
+-- Scritps
```
```suggestion
-- Scripts
```

Soupstraw commented on eras/conway/test-suite/test/Tests.hs:
```diff
@@ -4,6 +4,7 @@ module Main where
 
 import Cardano.Ledger.Conway (Conway)
 import Data.Proxy (Proxy (..))
+
```
Is there a reason for this newline?

lehins replied:
Nope. Just an accident

lehins replied:
Fixed

Soupstraw commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/CostModels.hs:
```diff
@@ -205,6 +206,9 @@ getCostModelLanguage (CostModel lang _ _) = lang
 getCostModelParams (CostModel _ cm _) = cm
 
+getCostModelEvaluationContext :: CostModel -> PV1.EvaluationContext
+getCostModelEvaluationContext = cmEvalCtx
```
Could we use the field accessor directly?

lehins replied:
Oh crap, thanks to your question I just found a bigger issue. `CostModel` constructor should not have been exported and the smart constructor `mkCostModel` must be used instead. That was why those accessor functions (`getCostModelLanguage` and `getCostModelParams`) where added in the first place.

So, the answer to your question is no. We should not be able to use the `cmEvalCtx` accessor directly.

It was accidentally exported in #3893. I've fixed it by hiding back the constructor.

Soupstraw commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/AlonzoInvalidTxUTXOW.hs:
```diff
@@ -925,7 +930,7 @@ noCostModelTx pf =
         , Outputs' [newTxOut pf [Address (someAddr pf), Amount (inject $ Coin 4995)]]
         , Txfee (Coin 5)
-        , WppHash (newScriptIntegrityHash pf (pp pf) [PlutusV2] redeemers (mkTxDats (Data (PV1.I 123))))
+        , WppHash (newScriptIntegrityHash pf pp' [PlutusV1] redeemers (mkTxDats (Data (PV1.I 123))))
```
Why was this changed to `PlutusV1`?

lehins replied:
Because `PlutusV2` is not available in Alonzo.

Soupstraw commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/AlonzoInvalidTxUTXOW.hs:
```diff
@@ -343,11 +344,14 @@ alonzoUTXOWTests pf =
+                  (initUTxO pf)
+                  pp'
+                  (trustMeP pf True $ noCostModelTx pf pp')
+                  ( Left [fromUtxos @era (CollectErrors [NoCostModel PlutusV1])]
```
Why was this changed to `PlutusV1`?

lehins replied:
Because PlutusV2 is not available in Alonzo.

Soupstraw commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/AlonzoCollectInputs.hs:
```diff
@@ -164,25 +165,26 @@ collectInputs ::
+mkPlutusScriptContext' (Alonzo _) = mkPlutusScriptContext
+mkPlutusScriptContext' (Babbage _) = mkPlutusScriptContext
+mkPlutusScriptContext' (Conway _) = mkPlutusScriptContext
+mkPlutusScriptContext' era = error ("mkPlutusScriptContext is not defined in era " ++ show era)
```
```suggestion
mkPlutusScriptContext' era = error ("mkPlutusScriptContext' is not defined in era " ++ show era)
```

lehins replied:
I've changed the error message to actually mean that the function `mkPlutusScriptContext` is not defined for that era. The fact that `mkPlutusScriptContext'` function is throwing an error will be apparent from the call stack. Which is why it usually makes no sense to add a name of the function into the error message (it was useful in the old days when `error` did not have `HasCallStack`).

Soupstraw commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/TxInfo.hs:
```diff
@@ -155,17 +134,12 @@ transCred (KeyHashObj (KeyHash kh)) =
-    -- The presence of a Byron address is caught above in the Just case
-    Nothing -> transAddr (txOut ^. addrTxOutL)
+transAddr addr = do
+  Addr _netId object stake <- Just addr
```
```suggestion
  let Addr _netId object stake = addr
```

Soupstraw replied:
Actually, can this function even return a `Nothing`? It should probably just return a `PV1.Address`

Soupstraw replied:
Ok, I see what it's doing now, but I'm not sure I prefer it over the old implementation

lehins replied:
Sure. Switched to an explicit case statement

Soupstraw commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/TxInfo.hs:
```diff
@@ -179,46 +153,17 @@ slotToPOSIXTime ei sysS s = do
-class Era era => EraPlutusContext (l :: Language) era where
-  transTxCert :: TxCert era -> PlutusTxCert l
-
-transWithdrawals :: Withdrawals c -> Map.Map PV1.StakingCredential Integer
-transWithdrawals (Withdrawals mp) = Map.foldlWithKey' accum Map.empty mp
-  where
-    accum ans (RewardAcnt _network cred) (Coin n) =
-      Map.insert (PV1.StakingHash (transCred cred)) n ans
-
-getWitVKeyHash :: (Crypto c, Typeable kr) => WitVKey kr c -> PV1.PubKeyHash
-getWitVKeyHash =
-  PV1.PubKeyHash
-    . PV1.toBuiltin
-    . hashToBytes
-    . (\(KeyHash x) -> x)
-    . hashKey
-    . (\(WitVKey x _) -> x)
```
Were these unused all this time?

lehins replied:
`getWitVKeyHash` was really dead code, unused anywhere

`transWithdrawals` have moved to `Alonzo.TxInfo`, since it  can only be used in Alonzo and Babbage and there is a new implementation of this function in Conway

Soupstraw commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/Language.hs:
```diff
@@ -63,46 +76,105 @@ import Data.Aeson (
-data Plutus = Plutus
-  { plutusLanguage :: !Language
-  , plutusScript :: !BinaryPlutus
+newtype PlutusRunnable (l :: Language) = PlutusRunnable
```
I think it'd help a lot to have docs for all of the new Plutus types, because it's a bit hard to keep track what they all do.

lehins replied:
Thank you for pointing out lack of haddock. I've added some. Please let me know if it can be improved.

Soupstraw replied:
Looks good to me

Soupstraw reviewed (commented):
Looks great so far. I'll finish the rest of the review tomorrow. Some of the comments I wrote down for myself as notes, so feel free to ignore them for now.

lehins committed:
Hide `CostModel` constructor.

Fix this mistake of exporting the `CostModel` constructor, which was introduced
in #3893.

lehins commented:
@Soupstraw I've addressed all of your comments.

Soupstraw commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Plutus/TxInfo.hs:
```diff
@@ -103,147 +73,188 @@ import Cardano.Ledger.Binary.Coders (
-    where
-      dec 0 = SumD ByronTxOutInContext <! From
-      dec 1 = SumD TranslationLogicMissingInput <! From
-      dec 2 = SumD RdmrPtrPointsToNothing <! From
-      dec 3 = SumD LanguageNotSupported <! From
-      dec 4 = SumD InlineDatumsNotSupported <! From
-      dec 5 = SumD ReferenceScriptsNotSupported <! From
-      dec 6 = SumD ReferenceInputsNotSupported <! From
```
We got rid of all of this with type level magic? :astonished:

lehins replied:
For Alonzo they are gone indeed. 

Those failures are all still present in Babbage and Conway, except for `LanguageNotSupported`, which was removed due to type level magic

Soupstraw commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs:
```diff
@@ -103,58 +154,108 @@ instance NoThunks Tag
+plutusScriptBinary :: AlonzoEraScript era => PlutusScript era -> PlutusBinary
+plutusScriptBinary ps = withPlutusScript ps plutusBinary
+
+-- | Verifies wether Plutus script is well formed or not, which simply means wether it
```
```suggestion
-- | Verifies whether Plutus script is well formed or not, which simply means whether it
```

Soupstraw commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs:
```diff
@@ -180,54 +281,63 @@ instance DecCBOR Tag where
+    Nothing ->
+      case getNativeScript script of
+        Just timelockScript -> deepseq timelockScript True
+        Nothing -> False -- Impossible case
```
Maybe we should throw an error here instead? That's how we've usually done it for impossible cases.

lehins replied:
Sure, I can switch it to `error`.

lehins replied:
Done.

Soupstraw commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxAuxData.hs:
```diff
@@ -136,37 +142,38 @@ instance Era era => EncCBOR (AlonzoTxAuxDataRaw era) where
-    partitionScripts (tss, pss1, pss2, pss3) =
+    partitionScripts (tss, pss) =
       \case
-        TimelockScript ts -> (ts :<| tss, pss1, pss2, pss3)
-        PlutusScript (Plutus PlutusV1 ps1) -> (tss, ps1 : pss1, pss2, pss3)
-        PlutusScript (Plutus PlutusV2 ps2) -> (tss, pss1, ps2 : pss2, pss3)
-        PlutusScript (Plutus PlutusV3 ps3) -> (tss, pss1, pss2, ps3 : pss3)
```
What the hell is this, Haskell for cats? :cat:

lehins replied:
:rofl:

Soupstraw commented on eras/alonzo/test-suite/src/Test/Cardano/Ledger/Alonzo/Translation/TranslatableGen.hs:
```diff
@@ -36,49 +49,65 @@ import Test.QuickCheck (
 import Test.QuickCheck.Gen (Gen (MkGen))
 import Test.QuickCheck.Random (mkQCGen)
 
+data TxInfoLanguage era where
```
I think this could also do with some haddock.

lehins replied:
I have a separate plan for PlutusContext golden tests, which will lead to this type disappearing, so I won't bother documenting it now.

lehins replied:
This will be done as part of this ticket: https://github.com/input-output-hk/cardano-ledger/issues/3923

Soupstraw commented on eras/conway/test-suite/test/Test/Cardano/Ledger/Conway/TxInfo.hs:
```diff
@@ -4,18 +4,12 @@
   Proxy era ->
   TestTree
-txInfoTests p = B.txInfoTestsV2 p PlutusV3 -- The V2 tests in Babbage should all hold for V3
+txInfoTests _p = testGroup "Conway Plutus Tests" [] -- Disabled for now
```
Make sure to have an issue to re-enable this at some point.

lehins replied:
Done: https://github.com/input-output-hk/cardano-ledger/issues/3923

Soupstraw commented on libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/Evaluate.hs:
```diff
@@ -10,296 +10,219 @@
-  , pwcDatums :: ![Data era]
-  , pwcExUnits :: !ExUnits
-  , pwcCostModel :: !CostModel
+data PlutusWithContext where
```
This needs some haddock

lehins replied:
Added some haddock

Soupstraw commented on libs/cardano-ledger-api/src/Cardano/Ledger/Api/Scripts/ExUnits.hs:
```diff
@@ -51,44 +38,55 @@ import Cardano.Ledger.Alonzo.TxWits (
     -- we cannot connect to a Plutus script.
     RedeemerPointsToUnknownScriptHash !RdmrPtr
-  | -- | Missing redeemer. The first parameter is the redeemer pointer which cannot be resolved,
+  | -- | Missing redeemer. The first parameter is the ,
```
the what?

lehins replied:
Thank you for catching this. I meant to remove the whole sentence, since I moved the documentation to the actual parameters (haddock nowadays supports this)

teodanciu commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Plutus/Evaluate.hs:
```diff
@@ -75,43 +57,35 @@ import NoThunks.Class (NoThunks)
   | NoWitness !(ScriptHash (EraCrypto era))
   | NoCostModel !Language
-  | BadTranslation !(TranslationError (EraCrypto era))
+  | BadTranslation !(ContextError era)
```
I see none of the changes (including this, exporting lookupPlutusScript, removing collectTwoPhaseScriptInputs ) are mentioned in the CHANGELOG, but perhaps they are not intended to be used by anyone else.
`CollectError` I see is used the Utxos rule predicate failure, so maybe it would help to mention this change?

lehins replied:
They absolutely have to be in the CHANGELOG. Thank you for reminding me to finish adjusting changelogs. I stopped half way through the PR.

lehins replied:
I've added a bunch of missing changelog entries for `cardano-ledger-alonzo`

Soupstraw commented on libs/cardano-ledger-api/src/Cardano/Ledger/Api/Scripts/ExUnits.hs:
```diff
@@ -51,44 +38,55 @@ import Cardano.Ledger.Alonzo.TxWits (
   | -- | Plutus evaluation error, for any version
-    ValidationFailure ValidationFailed
+    ValidationFailure
+      -- | Supplied execution units in the transactoin, which were ignored for calculating
```
```suggestion
      -- | Supplied execution units in the transaction, which were ignored for calculating
```

Soupstraw commented on libs/cardano-ledger-api/src/Cardano/Ledger/Api/Scripts/ExUnits.hs:
```diff
@@ -183,185 +172,81 @@ evalTxExUnitsWithLogs ::
+        note (NoCostModelInLedgerState lang) $ Map.lookup lang costModels
+      -- Similar to getSpendingDatum, but with more informative errors. It is OK to use
+      -- inline datums, when they are present, since for PlutusV1 presence of inline
+      -- datums would short circuit earlier on PlutuContext translation.
```
```suggestion
      -- datums would short circuit earlier on PlutusContext translation.
```

Soupstraw commented on eras/conway/impl/src/Cardano/Ledger/Conway/TxInfo.hs:
```diff
@@ -10,183 +11,301 @@
+  encCBOR = \case
+    -- We start at tag 8, just in case to avoid clashes with previous eras.
+    BabbageContextError babbageContextError ->
+      encode $ Sum BabbageContextError 8 !> To babbageContextError
```
Why are we encoding an additional tag here? Couldn't we just use the same tag as the final error that we're actually encoding?

lehins replied:
I don't understand your question. Could you make a code suggestion, so it is clear what you mean?

Soupstraw reviewed (approved):
This is a huge improvement. LGTM overall, but I'm sure I didn't catch everything because it's such a massive PR with lots of moving parts.

Soupstraw commented on eras/conway/impl/src/Cardano/Ledger/Conway/TxInfo.hs:
```diff
@@ -10,183 +11,301 @@
+  encCBOR = \case
+    -- We start at tag 8, just in case to avoid clashes with previous eras.
+    BabbageContextError babbageContextError ->
+      encode $ Sum BabbageContextError 8 !> To babbageContextError
```
Why not this?
```suggestion
      encCBOR babbageContextError
```

lehins replied:
Oh, I see what you mean now. We'll use `Sum` coders because there will be a bunch of errors added here when we finish integration of the Conway features (eg. VotingNotSupported for PlutusV1 and PlutusV2)

Soupstraw replied:
Babbage also uses `Sum` coders, so we should invoke those from Conway as well. Otherwise the encoding will have redundant tags that won't add any useful information (e.g. `InlineDatumsNotSupported` would be encoded as `\x08\x04...` instead of just `\x04...`). And we can still add new Conway errors from tag 8 and upwards.

Soupstraw replied:
I think that's how we do it for PredicateFailures also

lehins replied:
Those are not the types that are being stored on chain, so we do not need to worry about representing them in the most efficient manner, if that is your concern.

If I am missing what you are trying to say please elaborate. Or maybe produce a full DecCBOR instance of what you mean, because I don't understand the issue that you are trying to address here.

Soupstraw replied:
Yeah, I think we can leave it as is if we're not worried about efficiency.

teodanciu commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Plutus/Context.hs:
```diff
@@ -0,0 +1,125 @@
+import qualified PlutusLedgerApi.V3 as PV3
+
+class (PlutusLanguage l, AlonzoEraScript era) => EraPlutusTxInfo (l :: Language) era where
+  toPlutusTxCert :: proxy l -> TxCert era -> Either (ContextError era) (PlutusTxCert l)
```
I'm confused about the `ContextError` in these functions. Is it the `data  ContextError era :: Type` defined below in `EraPlutusContext era` class?  I suppose yes, but there needs to be nothing else specified to link the two in that case? Or is that happening when defining the instances? 
If the question doesn't make sense, please ignore it, I will ask you during the meeting tomorrow!

lehins replied:
I think I might understand the question. I can add `EraPlutusContext` as a superclass to the `EraPlutusTxInfo` if that would make it clearer. However, it is not necessary for a very interesting reason:

Associated data and type families (ones that are defined in a type class) are no different from the standalone data and type families and can in fact be treated the same at the use site. I think the only difference between the two is the fact that you can not create a type family instance without adding a type class instance for the associated type families.

teodanciu replied:
Interesting, thank you for the explanation! :+1:

lehins replied:
I did add `EraPlutusContext` as a superclass, which made a few constraints in one of the testing modules redundant :slightly_smiling_face:

teodanciu commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Plutus/Context.hs:
```diff
@@ -0,0 +1,125 @@
+  where
+  data ContextError era :: Type
+
+  mkPlutusScriptContext ::
```
So this function, ` mkPlutusScriptContext `, seems to be the "boss" . 
All the functions from `EraPlutusTxInfo` (  `toPlutusTxCert`, `toPlutusScriptPurpose`, `toPlutusTxInfo`, `toPlutusScriptContext`)   are used to build  `mkPlutusLanguageContext`, which in turn is used in each  `mkPlutusScriptContext ` implementation.  It is the one ultimately used in  in `Evaluate.collectPlutusScriptsWithContext` to build `PlutusWithContext`, which then gets passed to `evalPlutusScripts` in  Utxos rule. 
Not really a question or useful comment, sorry - just validating my understanding.  Anyway, such a clean model compared to `ExtendedUtxo` , `txInfo` and `valContext` - and their unclear (to me) interactions. 
Love how it all fits together so neatly.

lehins replied:
You are right on point!

teodanciu commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Plutus/Evaluate.hs:
```diff
@@ -0,0 +1,214 @@
+
+-- ===============================================================
+-- From the specification, Figure 8 "Scripts and their Arguments"
+-- ===============================================================
```
Out of curiosity, I checked the alonzo spec, and Figure 8 is now something else: 

![2023-12-12_17-14](https://github.com/input-output-hk/cardano-ledger/assets/2070003/cf630748-a05f-43ab-b06a-55d53bc2828e)

Figure 7 is "Scripts and their arguments", but don't know, this comment seems pretty unnecessary to me , since the implementation is quite different.
Here is Figure 7:

![2023-12-12_17-14_1](https://github.com/input-output-hk/cardano-ledger/assets/2070003/0224adfb-1c8a-4aa4-a442-7f03e27b1272)

lehins replied:
Looks like the comment has off-by-one error :slightly_smiling_face: 

It should say:
```
 From the specification, Figure 7 "Scripts and their Arguments"
```

I'll fix it my subsequent PR.

teodanciu reviewed (approved):
Epic! 
The relationship between everything involved needed to execute the scripts is now so clear, as is, supporting a new Plutus Version, and/or Era. 
I'm in an awe of this work!  :bow:

lehins commented:
@Soupstraw and @teodanciu Thank you for reviewing this PR! I know it was not a small task.
